### PR TITLE
gossip: fix TOCTOU race in discovery alive handling

### PR DESCRIPTION
**Pull request by Ady0333**
#### Type of change
- Bug fix

#### Description
This PR fixes a race condition in gossip discovery that could cause a peer to panic.

In handleAliveMessage(), membership state was checked under multiple read locks. If a member was purged concurrently between these checks, the code could reach a panic path even though this is a valid concurrent case.

The fix makes the membership check atomic and handles the edge case by re-learning the member instead of panicking.

#### Additional details
Tested locally with go test ./gossip/discovery/....
image

#### Release Note
Fixes a TOCTOU race in gossip discovery that could cause a peer to panic when a
member is concurrently purged while processing an alive message.

### CODE REVIEW

**Comment by Ady0333** [comment]
Hi @C0rWin @denyeart @yacovm,
This PR fixes a TOCTOU race in gossip discovery that could cause a peer panic under concurrent member purge.

Would appreciate your review when you have time.
 Thanks!

**Comment by C0rWin** [comment]
> Hi @C0rWin @denyeart @yacovm, This PR fixes a TOCTOU race in gossip discovery that could cause a peer panic under concurrent member purge.
> 
> Would appreciate your review when you have time. Thanks!

It would be beneficial if you could add a unit test that demonstrates how the bug manifests, so it's clear you actually fixed the issue, and no regression will be introduced in the future.

File: gossip/discovery/discovery_impl.go [hunk 1]
```diff
@@ -505,20 +505,18 @@ func (d *gossipDiscoveryImpl) handleAliveMessage(m *protoext.SignedGossipMessage
...
 		return
 	}
 
-	d.lock.RLock()
-	_, isAlive := d.aliveLastTS[string(pkiID)]
-	lastDeadTS, isDead := d.deadLastTS[string(pkiID)]
-	d.lock.RUnlock()
-
 	if !isAlive && !isDead {
-		d.logger.Panicf("Member %s is known but not found neither in alive nor in dead lastTS maps, isAlive=%v, isDead=%v", m.GetAliveMsg().Membership.Endpoint, isAlive, isDead)
```
Comment (C0rWin) [inline gossip/discovery/discovery_impl.go:521]: While I think we shouldn't have paniced here, I still think it would be great to see unit test added explaining the case.

**Comment by Ady0333** [comment]
Thanks @C0rWin for the suggestion. I’ve added a unit test that covers the concurrent purge case and verifies that handleAliveMessage() no longer panics.

The test passes locally with go test ./gossip/discovery/.... and 
go test ./gossip/discovery -run TestHandleAliveMessage_RelearnsMemberAfterConcurrentPurge

image
image

**Comment by C0rWin** [comment]
@Ady0333, first of all,l thanks for taking the time to suggest the fix. I have two concerns:

To start with, the test creates the inconsistent state directly rather than actually triggering a concurrent race. While this is acceptable for unit testing (race conditions are hard to reproduce reliably), it doesn't prove the race can occur in production. However, the code structure before the fix clearly shows the race window exists, so the fix is justified.

Having said that, I think that your fix correctly addresses a real race condition by consolidating map reads into a single atomic operation. Therefore, I am in favour of merging it.

**Review by C0rWin — APPROVED** [review]
